TypeSystem/Upper-lower-bounds/up_A01_t01.dart

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion We define the upper bound of two types T1 and T2 to be UP(T1,T2)
+/// as follows.
+/// - UP(X1 & B1, T2) =
+///   - T2 if X1 <: T2
+///   - otherwise X1 if T2 <: X1
+///   - otherwise UP(B1a, T2) where B1a is the greatest closure of B1 with
+///     respect to X1, as defined in inference.md.
+/// - UP(T1, X2 & B2) =
+///   - X2 if T1 <: X2
+///   - otherwise T1 if X2 <: T1
+///   - otherwise UP(T1, B2a) where B2a is the greatest closure of B2 with
+///     respect to X2, as defined in inference.md.
+///
+/// @description Check that upper bound of intersection types is calculated as
+/// expected
+/// @author sgrekhov22@gmail.com
+
+int? foo() => DateTime.now().millisecondsSinceEpoch.isEven? null : 1;
+
+void f<X extends num>(bool b, X x) {
+  if (x is int) { // `x` promoted to `X & int`.
+    var y1 = b ? x : null;
+    y1.isEven; // Error. Type of `y1` is `int?`
+//     ^^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
```

Comment:
The old (malformed) type was `(X & int)?`, so we should actually also test that the type is `int?` (not S?` for any other `S`). We could do that using `expectStaticType`.

`expectStaticType` wouldn't be quite safe, though, because it gives rise to erasure of intersection types (we can't pass `X & int` as a type argument to `expectStaticType`, explicitly or implicitly, so there's no way we can use `expectStaticType` to see the difference between `int?` and `(X & int)?)`. Actually, the reified type corresponding to `(X & int)?` isn't well-defined, because that type shouldn't exist in the first place.

In any case, we should also have something like this: `X? z = x;` and expect an error for that, to verify that the type of `x` is just `int?`, not `(X & int)?`.